Incorrect layout with large bitfield

Follow up to https://github.com/rust-lang-nursery/rust-bindgen/pull/1001

### Input C/C++ Header

```C++
struct {
  unsigned : 632;
} a;
```

### Bindgen Invocation

```
$ bindgen input.h
```

### Actual Results

Panic when running the layout tests: our generated struct ends up with the wrong size.

### Expected Results

We generate a struct with the correct layout, and it passes its layout tests.

@fitzgen so when the Item info for the bitfield is being printed out, it outputs this.
```
ir: ItemId(1) = Item {
    id: ItemId(
        1
    ),
    local_id: Cell {
        value: Some(
            1
        )
    },
    next_child_local_id: Cell {
        value: 1
    },
    canonical_name_cache: RefCell {
        value: None
    }, "correct" behavior for rust
    comment: None,
    annotations: Annotations {
        opaque: false,
        hide: false,
        use_instead_of: None,
        disallow_copy: false,
        private_fields: None,
        accessor_kind: None,
        constify_enum_variant: false
    },
    parent_id: ItemId(
        0
    ),
    kind: Type(
        Type {
            name: None,
            layout: Some(
                Layout {
                    size: 80,
                    align: 8,
                    packed: false
                }
            ),
            kind: Comp(
                CompInfo {
                    kind: Struct,
                    fields: AfterComputingBitfieldUnits(
                        [
                            Bitfields(
                                BitfieldUnit {
                                    nth: 1,
                                    layout: Layout {
                                        size: 128,
                                        align: 128,
                                        packed: false
                                    },
                                    bitfields: []
                                }
                            )
                        ]
                    ),
                    template_params: [],
                    methods: [],
                    constructors: [],
                    destructor: None,
                    base_members: [],
                    inner_types: [],
                    inner_vars: [],
                    has_own_virtual_method: false,
                    has_destructor: false,
                    has_nonempty_base: false,
                    has_non_type_template_params: false,
                    packed: false,
                    found_unknown_attr: false,
                    is_forward_declaration: false
                }
            ),
            is_const: false
        }
    )
}
```
So there are 2 different layouts being generated, one with size 80, align 8 and the other with size 128, align 128. Do you know where each of them are generated and what each of them means?

Another question I have is that when the header in question is compiled with clang++ it gives this warning
```
tests/headers/bitfield_large_overflow.hpp:2:12: warning: width of anonymous bit-field (632 bits) exceeds width of
      its type; value will be truncated to 32 bits [-Wbitfield-width]
  unsigned : 632;
           ^
```
If clang is truncating the value to 32 bits, does that mean the correct behavior would be to treat it as a 32 bit bitfield?

Here is the warning from gcc
```
tests/headers/bitfield_large_overflow.hpp:2:14: warning: width of ‘<anonymous struct>::<anonymous>’ exceeds its type
   unsigned : 632;
              ^
tests/headers/bitfield_large_overflow.hpp:3:3: warning: anonymous type with no linkage used to declare variable ‘<anonymous struct> a’ with linkage
 } a;
   ^
```
It doesn't explicitly say that it is truncating, so I am not 100% sure, but I think that is what it is doing. 
Another interesting behavior is that this only actually works with c++. Both gcc and clang fail to compile this code if the file is a .h, rather than just throwing an warning.
```
test.h:2:3: error: width of ‘<anonymous>’ exceeds its type
   unsigned : 632;
   ^
```

Looks like `g++` agrees with `clang++` on a size of `80`:

```
$ cat big-bitfield.cpp 
struct HasBigBitfield {
    unsigned : 632;
};

int f(HasBigBitfield*);

int main() {
    HasBigBitfield b;
    return f(&b);
}
$ g++ -g -c big-bitfield.cpp -o big-bitfield.o
big-bitfield.cpp:2:16: warning: width of ‘HasBigBitfield::<anonymous>’ exceeds its type
     unsigned : 632;
                ^~~
$ dwarfdump big-bitfield.o
...
< 1><0x0000002d>    DW_TAG_structure_type
                      DW_AT_name                  HasBigBitfield
                      DW_AT_byte_size             0x00000050
                      DW_AT_decl_file             0x00000001 /home/fitzgen/scratch/big-bitfield.cpp
                      DW_AT_decl_line             0x00000001
...
$ 
```

@fitzgen Another interesting thing I just found, the actual type assigned to the anonymous bitfield doesn't change the size of the struct.

```
$ cat big-bitfield.cpp 
struct HasBigBitfield {
    unsigned char: 632;
};

int f(HasBigBitfield*);

int main() {
    HasBigBitfield b;
    return f(&b);
}
$ clang++ -g -c big-bitfield.cpp -o big-bitfield.o
big-bitfield.cpp:2:19: warning: width of anonymous bit-field (632 bits) exceeds width of its type; value will be
      truncated to 8 bits [-Wbitfield-width]
    unsigned char : 632;
                  ^
1 warning generated.
$ dwarfdump big-bitfield.o
...
< 1><0x0000002a>    DW_TAG_structure_type
                      DW_AT_name                  "HasBigBitfield"
                      DW_AT_byte_size             0x00000050
                      DW_AT_decl_file             0x00000001 /home/oliver/rust-bindgen/big-bitfield.cpp
                      DW_AT_decl_line             0x00000001
...
$ 
```
g++ also has the same behavior. Considering that it specifically says that the value is being truncated, I find it odd that the memory is being allocated to fit the entire bitfield. 

I tried a few different combinations of bitfield sizes are the memory size pattern seems to be 0x40, 0x48, 0x50, 0x58, etc. Is it possible to just check whatever memory size libclang outputs and use that as the memory size? But does that mean we are just allocating extra memory for no reason?

--- a/src/codegen/impl_debug.rs
+++ b/src/codegen/impl_debug.rs
@@ -95,11 +95,14 @@ impl<'a> ImplDebug<'a> for BitfieldUnit {
             if i > 0 {
                 format_string.push_str(", ");
             }
-            format_string.push_str(&format!("{} : {{:?}}", bu.name()));
-            let name_ident = ctx.rust_ident_raw(bu.name());
-            tokens.push(quote! {
-                self.#name_ident ()
-            });
+
+            if let Some(name) = bu.name() {
+                format_string.push_str(&format!("{} : {{:?}}", name));
+                let name_ident = ctx.rust_ident_raw(name);
+                tokens.push(quote! {
+                    self.#name_ident ()
+                });
+            }
         }
 
         Some((format_string, tokens))

--- a/src/codegen/impl_partialeq.rs
+++ b/src/codegen/impl_partialeq.rs
@@ -51,10 +51,12 @@ pub fn gen_partialeq_impl(
                     tokens.push(gen_field(ctx, ty_item, name));
                 }
                 Field::Bitfields(ref bu) => for bitfield in bu.bitfields() {
-                    let name_ident = ctx.rust_ident_raw(bitfield.name());
-                    tokens.push(quote! {
-                        self.#name_ident () == other.#name_ident ()
-                    });
+                    if let Some(name) = bitfield.name() {
+                        let name_ident = ctx.rust_ident_raw(name);
+                        tokens.push(quote! {
+                            self.#name_ident () == other.#name_ident ()
+                        });
+                    }
                 },
             }
         }

--- a/src/codegen/mod.rs
+++ b/src/codegen/mod.rs
@@ -1185,6 +1185,10 @@ impl<'a> FieldCodegen<'a> for BitfieldUnit {
         let mut ctor_impl = quote! { 0 };
 
         for bf in self.bitfields() {
+            // Codegen not allowed for anonymous bitfields
+            if bf.name().is_none() {
+                continue;
+            }
             bf.codegen(
                 ctx,
                 fields_should_be_private,
@@ -1198,7 +1202,7 @@ impl<'a> FieldCodegen<'a> for BitfieldUnit {
                 (&unit_field_name, unit_field_int_ty.clone()),
             );
 
-            let param_name = bitfield_getter_name(ctx, parent, bf.name());
+            let param_name = bitfield_getter_name(ctx, parent, bf.name().unwrap());
             let bitfield_ty_item = ctx.resolve_item(bf.ty());
             let bitfield_ty = bitfield_ty_item.expect_type();
             let bitfield_ty =
@@ -1307,9 +1311,11 @@ impl<'a> FieldCodegen<'a> for Bitfield {
         F: Extend<quote::Tokens>,
         M: Extend<quote::Tokens>,
     {
+        // Should never be called with name() as None, as codegen can't be done
+        // on an anonymous bitfield
         let prefix = ctx.trait_prefix();
-        let getter_name = bitfield_getter_name(ctx, parent, self.name());
-        let setter_name = bitfield_setter_name(ctx, parent, self.name());
+        let getter_name = bitfield_getter_name(ctx, parent, self.name().unwrap());
+        let setter_name = bitfield_setter_name(ctx, parent, self.name().unwrap());
         let unit_field_ident = quote::Ident::new(unit_field_name);
 
         let bitfield_ty_item = ctx.resolve_item(self.ty());

--- a/src/ir/comp.rs
+++ b/src/ir/comp.rs
@@ -276,7 +276,7 @@ impl DotAttributes for Bitfield {
         writeln!(
             out,
             "<tr><td>{} : {}</td><td>{:?}</td></tr>",
-            self.name(),
+            self.name().unwrap_or("(anonymous)"),
             self.width(),
             self.ty()
         )
@@ -298,7 +298,6 @@ impl Bitfield {
     /// Construct a new bitfield.
     fn new(offset_into_unit: usize, raw: RawField) -> Bitfield {
         assert!(raw.bitfield().is_some());
-        assert!(raw.name().is_some());
 
         Bitfield {
             offset_into_unit: offset_into_unit,
@@ -331,11 +330,6 @@ impl Bitfield {
     /// Get the bit width of this bitfield.
     pub fn width(&self) -> u32 {
         self.data.bitfield().unwrap()
-    }
-
-    /// Get the name of this bitfield.
-    pub fn name(&self) -> &str {
-        self.data.name().unwrap()
     }
 }
 
@@ -581,13 +575,12 @@ fn bitfields_to_allocation_units<E, I>(
             }
         }
 
-        // Only keep named bitfields around. Unnamed bitfields (with > 0
-        // bitsize) are used for padding. Because the `Bitfield` struct stores
-        // the bit-offset into its allocation unit where its bits begin, we
-        // don't need any padding bits hereafter.
-        if bitfield.name().is_some() {
-            bitfields_in_unit.push(Bitfield::new(offset, bitfield));
-        }
+        // Always keep all bitfields around. While unnamed bitifields are used
+        // for padding (and usually not needed hereafter), large unnamed
+        // bitfields over their types size cause weird allocation size behavior from clang.
+        // Therefore, all bitfields needed to be kept around in order to check for this
+        // and make the struct opaque in this case
+        bitfields_in_unit.push(Bitfield::new(offset, bitfield));
 
         max_align = cmp::max(max_align, bitfield_align);
 

--- a/tests/expectations/tests/bitfield_large_overflow.rs
+++ b/tests/expectations/tests/bitfield_large_overflow.rs
@@ -18,4 +18,3 @@ extern "C" {
     #[link_name = "a"]
     pub static mut a: _bindgen_ty_1;
 }
-

